File: Objective C/testTek/ViewController.m
//
//  ViewController.m
//  testTek
//
//  Created by applecaca on 09/09/15.
//  Copyright © 2015 stupeflix. All rights reserved.
//

#import "ViewController.h"
#import "DataController.h"
#import "SFTableView.h"
#import "SFTableViewCell.h"
#import "SFTableViewCell.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;
@property (nonatomic, strong) NSArray *employeList;
@property (weak, nonatomic) IBOutlet SFTableView *tableView;


@end

@implementation ViewController

#pragma mark View life Cycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self.spinner setHidden:NO];
    [self.spinner startAnimating];
    
    [[DataController sharedInstance] fetchJsonWithCompletionBlock:^(NSArray *pemployeList, NSError *error)
    {
        if (pemployeList && error == nil)
        {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.employeList = pemployeList;
                [self.spinner stopAnimating];
                [self.spinner setHidden:YES];
                self.tableView.delegate = self;
            });
        }
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark SFTableViewDelegate

- (SFTableViewCell *)tableView:(SFTableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.employeList != nil)
    {
        return [SFTableViewCell cellWithEmploye:self.employeList[indexPath.row]];
    }
    return nil;
}

- (NSInteger)tableView:(SFTableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (self.employeList == nil) return 0;
    
    return self.employeList.count;
}

@end

File: Swift/testTekSwift/testTekSwift/ViewController.swift
//
//  ViewController.swift
//  testTekSwift
//
//  Created by Nicolas Bellon on 27/10/2015.
//  Copyright © 2015 daid. All rights reserved.
//

import UIKit

class ViewController: UIViewController, SFTableViewDataSource
{
    private var TableView : SFTableView;
    private var employeList : [Employe]?;
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?)
    {
        self.TableView = SFTableView.init(frame: UIScreen.mainScreen().bounds);
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil);
    }

    required init?(coder aDecoder: NSCoder)
    {
        self.TableView = SFTableView.init(frame: UIScreen.mainScreen().bounds);
        super.init(coder: aDecoder);
    }
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        self.view.addSubview(self.TableView);
        DataController.sharedInstance.fetchEmployeListFromJson { (employe: [Employe]?, error: NSError?) -> Void in
            
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                if employe != nil
                {
                    self.employeList = employe;
                    self.TableView.dataSource = self;
                }
                else
                {
                    let unpackError = error!;
                    
                    var message = "Impossible de télecharger les données";
                    
                    let faillureReason : String? = unpackError.userInfo[NSLocalizedFailureReasonErrorKey] as! String?;
                    
                    if faillureReason != nil
                    {
                        message = faillureReason!;
                    }
                    
                    let alert : UIAlertController = UIAlertController.init(title: "Test Tek Swift", message: message, preferredStyle: UIAlertControllerStyle.Alert)
                    let alertAction : UIAlertAction = UIAlertAction.init(title: "OK", style: UIAlertActionStyle.Default, handler: nil);
                    alert.addAction(alertAction);
                    self.presentViewController(alert, animated: true, completion: nil);
                }

            });
            
        }
    }

    
    func tableView(tableView: SFTableView, numberOfRowsInSection section: Int) -> Int
    {
        if let employeList = employeList
        {
            return employeList.count;
        }
        
        return 0;
    }
    
    func tableView(tableView: SFTableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> SFTableViewCell
    {
        
        if let employeList = employeList
        {
            let emp : Employe = employeList[indexPath.row];
            let cell : SFTableViewCell? = SFTableViewCell.initWithEmploye(emp);

            if let cell = cell
            {
                return cell;
            }
        }
        
        return SFTableViewCell.init(frame: CGRectZero);
    }
}

